typegen/src/utils.rs

```diff
@@ -0,0 +1,130 @@
+    let mut renaming_commands: Vec<(u32, usize)> = vec![];
+
+    for type_ids in clashing_type_ids {
+        // Map N types in type_ids to M new type paths, where M <= N.
+        // types with the same structure should map to the same type path after all.
+        let mut types_with_same_structure: Vec<(&Type<PortableForm>, SmallVec<[u32; 2]>)> = vec![];
+        'outer: for id in type_ids {
+            let type_a = types.resolve(id).expect("type is present; qed;");
+            for (type_b, ids) in types_with_same_structure.iter_mut() {
+                if types_equal_extended_to_params(type_a, type_b) {
+                    ids.push(id);
+                    continue 'outer;
+                }
+            }
+            types_with_same_structure.push((type_a, smallvec![id]));
+        }
+
+        // Now that the types that share a structure are grouped together, we can order commands to rename them.
+        for (n, (_, group)) in types_with_same_structure.iter().enumerate() {
+            for id in group {
+                renaming_commands.push((*id, n + 1));
+            }
+        }
+    }
+
+    // execute the actual renaming. The `get_mut()` with the usize cast is a bit awkward, but there is currently not `resolve_mut()` function on the `PortableRegistry`.
+    for (id, appendix) in renaming_commands {
+        let ty = types
+            .types
+            .get_mut(id as usize)
+            .expect("type is present; qed;");
+        assert_eq!(ty.id, id);
+        let name = ty.ty.path.segments.last_mut().expect("This is only empty for builtin types, that are filtered out with namespace().is_empty() above; qed;");
+        *name = format!("{name}{appendix}"); // e.g. Header1, Header2, Header3, ...
+    }
+}
+
+/// This function checks if two types that have the same type path,
+/// should be considered as equal in terms of their structure and
+/// their use of generics.
+/// This is checked, because if they are indeed equal it is fine
+/// to generate a single rust type for the two registered typed.
+/// However if they are not equal, we need to deduplicate the type path.
+/// This means modifying the type path of one or both clashing types.
+///
+/// So what types are considered equal?
+/// Types are equal, if their TypeDef is exactly the same.
+/// Types are also considered equal if the TypeDef has the same shape and
+/// all type ids mentioned in the TypeDef are either:
+/// - equal
+/// - or different, but map essentially to the same generic type parameter
+fn types_equal_extended_to_params(a: &Type<PortableForm>, b: &Type<PortableForm>) -> bool {
+    let collect_params = |type_params: &[TypeParameter<PortableForm>]| {
+        type_params
+            .iter()
+            .filter_map(|p| p.ty.map(|ty| (ty.id, p.name.clone())))
+            .collect::<HashMap<u32, String>>()
+    };
+
+    let type_params_a = collect_params(&a.type_params);
```

Comment:
I geuss there's a possibility that two types at the same path can be different in terms of how many type params they have too, so we should probably check that the type param count/names are identical too?

And then if it's identical, and the shape is the same, then IDs are equal as long as they are the same, or they point to a generic at the same _index_ (could compare names too ofc, but index would work too and be cheap)